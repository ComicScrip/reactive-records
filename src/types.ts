import { Record, Collection } from "./internals"
import { Scope } from "./Scope"

export type Partial<T> = { [P in keyof T]?: T[P] }
export type OptimisticPrimaryKey = string
export type PrimaryKey = string | number

export interface RecordConstructor {
  new (): Record
}

export interface toOneAssociationDescription<T> {
  foreignCollection: () => Collection<Record>
  foreignKeyAttribute?: keyof T
}

export interface toOneAssociationsDescription<T> {
  [name: string]: toOneAssociationDescription<T>
}

export interface toManyAssociationDescription<T> {
  foreignCollection: () => Collection<Record>
  foreignKeyAttribute?: keyof T
}

export interface toManyAssociationsDescription<T> {
  [name: string]: toManyAssociationDescription<T>
}

export type PersistenceServiceName = string

export interface PersistenceStrategy {
  persistenceServices: object
  loadMany(params: any, scope: Scope<Record>): Promise<any>
  loadOne(params: any, record: Record, scope: Scope<Record>): Promise<any>
  saveOne(params: any, record: Record, scope: Scope<Record>): Promise<any>
  destroyOne(params: any, record: Record, scope: Scope<Record>): Promise<any>
}
