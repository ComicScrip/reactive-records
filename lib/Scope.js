"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var mobx_1 = require("mobx");
var Bluebird = require("bluebird");
Promise = Bluebird;
/**
 * A subset of a collection
 */
var Scope = /** @class */ (function () {
    function Scope(collection, name, params) {
        if (name === void 0) { name = "default"; }
        if (params === void 0) { params = {}; }
        /**
         * Holds an ordered list of records primary keys
         */
        this.itemPrimaryKeys = [];
        /**
         * Either false or the persistence service currently used to fetch the items
         * You can use this for exemple to differentiate data loaded form local app storage or a remote API
         */
        this.loadingFrom = false;
        /**
         * Either false or the persistence service lastly used to fetch the items
         * You can use this for exemple to differentiate data loaded form local app storage or a remote API
         */
        this.lastLoadedFrom = false;
        /**
         * Either null or the time when the scope was lastely loaded
         * You can use this for exemple to decide if a scope can be returned from cache when you have to (re)load it
         */
        this.lastLoadedAt = null;
        this.params = params;
        this.name = name;
        this.collection = collection;
    }
    /**
     * Sets the scope's paramaters
     * @param {object} params
     * @return {this<RecordType extends Record>}
     */
    Scope.prototype.setParams = function (params) {
        this.params = params;
        return this;
    };
    /**
     * Load items for this scope into the scope's collection
     * @param {object} params
     * @return {Promise<any>}
     */
    Scope.prototype.load = function (params) {
        return this.collection.load(params ? params : this.params, this.name);
    };
    /**
     * returns true if the scope contains the pk
     */
    Scope.prototype.hasPk = function (pk) {
        return this.itemPrimaryKeys.findIndex(function (p) { return p == pk; }) !== -1;
    };
    /**
     * adds primary key in the scope's pk list, if the latter is not already there
     */
    Scope.prototype.addPk = function (pk) {
        if (!this.hasPk(pk)) {
            this.itemPrimaryKeys.push(pk);
        }
        return this;
    };
    /**
     * replace primary key in the scope's pk list by another one
     */
    Scope.prototype.replacePk = function (oldpk, newPk) {
        var pkIndex = this.itemPrimaryKeys.findIndex(function (p) { return p == oldpk; });
        if (pkIndex !== -1) {
            this.itemPrimaryKeys[pkIndex] = newPk;
        }
        return this;
    };
    /**
     * Removes a pk form the set
     * @param {PrimaryKey} pk
     * @return {Scope<RecordType extends Record>}
     */
    Scope.prototype.removePk = function (pk) {
        var idx = this.itemPrimaryKeys.findIndex(function (p) { return p == pk; });
        if (idx !== -1) {
            this.itemPrimaryKeys.splice(idx, 1);
        }
        return this;
    };
    Object.defineProperty(Scope.prototype, "items", {
        /**
         * get the scope items instances
         */
        get: function () {
            var result = [];
            for (var i = 0; i < this.itemPrimaryKeys.length; i++) {
                var item = this.collection.get(this.itemPrimaryKeys[i]);
                if (item) {
                    result.push(item);
                }
            }
            return result;
        },
        enumerable: true,
        configurable: true
    });
    __decorate([
        mobx_1.observable
    ], Scope.prototype, "itemPrimaryKeys", void 0);
    __decorate([
        mobx_1.observable
    ], Scope.prototype, "loadingFrom", void 0);
    __decorate([
        mobx_1.observable
    ], Scope.prototype, "lastLoadedFrom", void 0);
    __decorate([
        mobx_1.observable
    ], Scope.prototype, "lastLoadedAt", void 0);
    __decorate([
        mobx_1.action.bound
    ], Scope.prototype, "setParams", null);
    __decorate([
        mobx_1.action.bound
    ], Scope.prototype, "load", null);
    __decorate([
        mobx_1.action.bound
    ], Scope.prototype, "addPk", null);
    __decorate([
        mobx_1.action.bound
    ], Scope.prototype, "replacePk", null);
    __decorate([
        mobx_1.action.bound
    ], Scope.prototype, "removePk", null);
    __decorate([
        mobx_1.computed
    ], Scope.prototype, "items", null);
    return Scope;
}());
exports.Scope = Scope;
