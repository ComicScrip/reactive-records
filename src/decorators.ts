import { isArray, isObject, isEmpty } from "lodash"
import {
  Record,
  toOneAssociationDescription,
  toManyAssociationDescription
} from "./internals"

/**
 * Marks a Record member as being an ownAttribute of a Record
 * @param target The record
 * @param key The member to be marked as ownAttribute
 */
export function ownAttribute(target: Record, key: string) {
  const constructor = target.constructor as any
  if (!isArray(constructor._ownAttributeNames)) {
    constructor._ownAttributeNames = []
  }
  constructor._ownAttributeNames.push(key)
}

/**
 * Returns a decorator function that marks a Record class member as being a 'toOneAssociation' association of a Record
 * @param associationDescription The description object provided for the association :
 * *foreignCollection*: the associated Record(s)'s foreignCollection instance
 * *foreignKeyAttribute* the name of the property that contains the associated Record(s)'s primary key(s)
 */
export function toOneAssociation<T>(
  associationDescription: toOneAssociationDescription<T>
) {
  return (target: Record, key: string) => {
    const constructor = target.constructor as any
    if (!isObject(constructor._toOneAssociations)) {
      constructor._toOneAssociations = {}
    }
    constructor._toOneAssociations[key] = associationDescription
  }
}

/**
 * Returns a decorator function that marks a Record class member as being a 'toManyAssociation' association of a Record
 * @param associationDescription The description object provided for the association :
 * *foreignCollection*: the associated Record(s)'s foreignCollection instance
 * *selfReferenceAttribute* the name of the property of the foreign records that may contains this Record(s)'s primary key(s)
 */
export function toManyAssociation<T>(
  associationDescription: toManyAssociationDescription<T>
) {
  return (target: Record, key: string) => {
    const constructor = target.constructor as any
    if (!isObject(constructor._toManyAssociations)) {
      constructor._toManyAssociations = {}
    }
    constructor._toManyAssociations[key] = associationDescription
  }
}
