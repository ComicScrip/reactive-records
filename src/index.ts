import {
  Collection,
  Record,
  ownAttribute,
  toOneAssociation,
  PrimaryKey,
  Partial,
  toOneAssociationDescription,
  toOneAssociationsDescription,
  toManyAssociationDescription,
  toManyAssociationsDescription,
  PersistenceServiceName,
  PersistenceStrategy
} from "./internals"

export {
  Collection,
  Record,
  ownAttribute,
  toOneAssociation,
  PrimaryKey,
  Partial,
  toOneAssociationsDescription,
  toManyAssociationDescription,
  toOneAssociationDescription,
  toManyAssociationsDescription,
  PersistenceStrategy,
  PersistenceServiceName
}
