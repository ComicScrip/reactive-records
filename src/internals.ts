export * from "./Collection"
export * from "./Record"
export * from "./types"
export * from "./decorators"
