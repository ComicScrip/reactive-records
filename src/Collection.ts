import { action, computed, observable, ObservableMap } from "mobx"
import { createTransformer } from "mobx-utils"
import { Record, Partial, PrimaryKey, PersistenceStrategy } from "./internals"
import { Scope } from "./Scope"

/**
 * A Store for records.
 * All Storage related manipulations are done here (local and remote synchronisations, instances in memory, ...)
 */
export abstract class Collection<RecordType extends Record> {
  /**
   * Holds all the collection's records' instances
   * The map is indexed by the records' primary key values.
   */
  @observable
  private records: ObservableMap<PrimaryKey, RecordType> = observable.map()

  /**
   * The collection's scopes
   */
  @observable
  private scopes: ObservableMap<string, Scope<RecordType>> = observable.map()

  public persistenceStrategy: PersistenceStrategy = null

  /**
   * Get the collection's records' constructor
   */
  public abstract get recordClass(): typeof Record

  /**
   * Returns the persistence strategy for this collection, or throw an error is not defined
   * @return {PersistenceStrategy}
   */
  public getPersistenceStrategy(): PersistenceStrategy {
    if (this.persistenceStrategy === null) {
      throw new Error(
        `Please define a persistence strategy for the collection '${
          (this.constructor as any).name
        }'`
      )
    }
    return this.persistenceStrategy
  }

  /**
   * Get a collection scope by name
   * @param {string} name : the name of the scope
   * @return {Scope<RecordType extends Record>}
   */
  public getScope(name: string): Scope<RecordType> | undefined {
    return this.scopes.get(name)
  }

  /**
   * Returns all the collection scopes names in an array
   * @return {string[]}
   */
  public get scopesNames(): string[] {
    return (this.scopes as any)._keys
  }

  /**
   * Get a collection scopes whose name matches a given regex
   * @return {Scope<RecordType extends Record>}
   * @param regex The regex that will be used against all scopes name
   */
  public getScopesMatching(regex: RegExp): Array<Scope<RecordType>> {
    return this.scopesNames
      .filter(scopeName => scopeName.match(regex))
      .map(name => this.scopes.get(name))
  }

  /**
   * Filters all the collection scopes with a regex and return their items concatenated
   */
  public getCombinedScopeItems = ((regex: RegExp) => {
    const scopeNames = this.scopesNames
    let items = []
    for (let i = 0; i < scopeNames.length; i++) {
      const scope = scopeNames[i]
      if (scope.match(regex)) {
        items = items.concat(this.scopes.get(scope).items)
      }
    }
    return items
  }).bind(this)

  /**
   * Takes all scopes matching a regex and concat their items
   * @return {RecordType[]}
   * @param regex The regex that will be used against all scopes name
   */
  public combineScopeItems = createTransformer(this.getCombinedScopeItems) as (
    regex: RegExp
  ) => RecordType[]

  /**
   * Get an existing scope or create a new one
   * @param {string} name : The name of the scope
   * @param {object} params : the scope's params to be set if specified
   * @return {Scope<RecordType extends Record>}
   */
  public provideScope(name: string = "default", params?: object): Scope<RecordType> {
    let scope = this.getScope(name)

    if (scope) {
      return scope
    } else {
      scope = new Scope(this, name, params)
      this.setScope(scope)
    }

    if (params) {
      scope.setParams(params)
    }

    return scope
  }

  /**
   * Set a scope into the collection's scope set
   * @param {Scope<RecordType extends Record>} scope
   */
  @action.bound
  public setScope(scope: Scope<RecordType>) {
    this.scopes.set(scope.name, scope)
  }

  /**
   * remove a scope from the collection
   * @param {Scope<RecordType extends Record>} scope
   */
  @action.bound
  public unsetScope(scope: Scope<RecordType>) {
    this.scopes.delete(scope.name)
  }

  /**
   * Set the collection's persistence strategy
   * @param ps {PersistenceStrategy} : The persistence strategy to set
   * @return {Collection<Record>} : The collection
   */
  @action.bound
  public setPersistenceStratgy(ps: PersistenceStrategy) {
    this.persistenceStrategy = ps
    return this
  }

  /**
   * Get an array of the collection's record's primary keys
   * @return {PrimaryKey[]}
   */
  @computed
  get itemsPrimaryKeys(): PrimaryKey[] {
    // ObservebaleMap.prototype._keys is much faster than keys() and returns directly what we need
    return (this.records as any)._keys
  }

  /**
   * Get a list of the _collection's record instances
   * @returns An array of Record instances
   */
  @computed
  public get items(): Array<RecordType> {
    const items = []
    for (let i = 0; i < this.itemsPrimaryKeys.length; i++) {
      const itemPrimaryKey = this.itemsPrimaryKeys[i]
      items[i] = this.records.get(itemPrimaryKey)
    }
    return items
  }

  /**
   * Indicates whether or not the _collection contains a record with a given primary key
   * @param primaryKey The primary key whose existence will be checked
   * @returns true if the primary key is present, false otherwise
   */
  public has(primaryKey): boolean {
    return this.records.has(primaryKey)
  }

  /**
   * Get a record instance in the _collection form its primary key
   * @param primaryKey the reocrd's primary key value
   * @return The record instance or undefined if there is no record with the given primary key
   */
  public get(primaryKey: PrimaryKey): RecordType {
    return this.records.get(primaryKey)
  }

  /**
   * Return the collection items filtered by a value on a property
   * @param {keyof RecordType} propName
   * @param propValue
   * @return {RecordType[]}
   */
  public wherePropEq(propName: keyof RecordType, propValue: any): RecordType[] {
    const filteredRecords = []
    for (let i = 0; i < this.items.length; i++) {
      const item = this.items[i]
      if (item[propName] == propValue) {
        filteredRecords.push(item)
      }
    }
    return filteredRecords
  }

  /**
   * Get multiple record instances in the _collection form their primary keys
   * @param primaryKeyList the reocrds' primary key values
   */
  public getMany(primaryKeyList: PrimaryKey[]): RecordType[] {
    const recordInstances = []
    for (let i = 0; i < primaryKeyList.length; i++) {
      const primaryKey = primaryKeyList[i]
      recordInstances.push(this.get(primaryKey))
    }
    return recordInstances
  }

  /**
   * Get the size of the collection
   * @returns The number of (all) items in the collection
   */
  @computed
  public get size(): number {
    return this.records.size
  }

  /**
   * Add or replace one record in the collection
   * @param recordProperties A plain object reprsentation of the record's properties
   */
  @action.bound
  public set(recordProperties: Partial<RecordType>): RecordType {
    // TODO: add an option to replace or not
    if (recordProperties instanceof Record) {
      return recordProperties as RecordType
    }
    const recordClass = this.recordClass
    const recordInstance = new recordClass(this) as RecordType
    recordInstance._mergeProperties(recordProperties)
    this.records.set(recordInstance._primaryKeyValue, recordInstance)

    return recordInstance
  }

  /**
   * Add or replace one record in the collection
   * @param recordInstance The record instance to set in the collection
   */
  @action.bound
  public setRecord(recordInstance: RecordType): RecordType {
    this.records.set(recordInstance._primaryKeyValue, recordInstance)
    return recordInstance
  }

  /**
   * Updates the index of the a record with a primaryKeyValue of oldPk (if found)
   * in the collection's records' map
   * @param oldPk the index to be replaced
   * @param newPk the new value of the index
   */
  @action.bound
  public updateRecordPrimaryKey(oldPk: PrimaryKey, newPk: PrimaryKey) {
    if (oldPk === newPk) return
    const record = this.get(oldPk)
    if (record) {
      this.records.set(newPk, this.get(oldPk))
      this.records.delete(oldPk)
    }
  }

  /**
   * Add or replace multiple records in the _collection
   * @param recordPropertiesList An array of plain object reprsentation of the records' properties
   */
  @action.bound
  public setMany(recordPropertiesList: Partial<RecordType>[]): RecordType[] {
    const recordInstances = []
    for (let i = 0; i < recordPropertiesList.length; i++) {
      const recordProperties = recordPropertiesList[i]
      recordInstances.push(this.set(recordProperties))
    }
    return recordInstances
  }

  /**
   * Delete a record with a given primary key in the _collection
   * @param primaryKey The identifier of the record to delete from the _collection
   */
  @action.bound
  public unset(primaryKey: PrimaryKey) {
    this.records.delete(primaryKey)
    return this
  }

  /**
   * Delete multipe records in the _collection form their primary keys
   * @param primaryKeyList The identifiers of the records to delete from the _collection
   */
  @action.bound
  public unsetMany(primaryKeyList: PrimaryKey[]) {
    for (let i = 0; i < primaryKeyList.length; i++) {
      const primaryKey = primaryKeyList[i]
      this.unset(primaryKey)
    }
    return this
  }

  /**
   * Delete all the records in the _collection
   */
  @action.bound
  public clear(): this {
    this.records.clear()
    return this
  }

  /**
   * Loads items into the collection using the collection's persitence strategy
   */
  @action.bound
  public async load(params: any = {}, scopeName: string = "default"): Promise<any> {
    return this.getPersistenceStrategy().loadMany(params, this.provideScope(scopeName, params))
  }

  /**
   * Loads a record into the collection using the collection's persitence strategy
   */
  @action.bound
  public async loadOne(
    record: Record | PrimaryKey,
    params: any = {},
    scopeName: string = "default"
  ): Promise<any> {
    if (!(record instanceof Record)) {
      const r = { [this.recordClass.primaryKeyName]: record } as any
      record = this.set(r)
    }
    return this.getPersistenceStrategy().loadOne(
      params,
      record as Record,
      this.provideScope(scopeName, params)
    )
  }

  /**
   * Saves one record into the collection using the collection's persitence strategy
   */
  @action.bound
  public async saveOne(
    record: Record,
    params: any = {},
    scopeName: string = "default"
  ): Promise<any> {
    return this.getPersistenceStrategy().saveOne(
      params,
      record,
      this.provideScope(scopeName, params)
    )
  }

  /**
   * Destroys one record of the collection using the collection's persitence strategy
   */
  @action.bound
  public async destroyOne(
    record: Record,
    params: any = {},
    scopeName: string = "default"
  ): Promise<any> {
    return this.getPersistenceStrategy().destroyOne(
      params,
      record,
      this.provideScope(scopeName, params)
    )
  }

  /**
   * Unset all scopes and records from the collection
   */
  @action.bound
  reset() {
    this.scopes = observable.map()
    this.clear()
  }
}
