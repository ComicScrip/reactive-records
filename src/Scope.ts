import { action, computed, observable } from "mobx"
import * as Bluebird from "bluebird"
Promise = Bluebird as any
import { Record, Partial, PrimaryKey, Collection, PersistenceServiceName } from "./internals"

/**
 * A subset of a collection
 */
export class Scope<RecordType extends Record> {
  /**
   * Holds an ordered list of records primary keys
   */
  @observable
  public itemPrimaryKeys: PrimaryKey[] = []

  /**
   * Either false or the persistence service currently used to fetch the items
   * You can use this for exemple to differentiate data loaded form local app storage or a remote API
   */
  @observable
  public loadingFrom: PersistenceServiceName | boolean = false

  /**
   * Either false or the persistence service lastly used to fetch the items
   * You can use this for exemple to differentiate data loaded form local app storage or a remote API
   */
  @observable
  public lastLoadedFrom: PersistenceServiceName | boolean = false

  /**
   * Either false or the persistence service lastly used to fetch the items
   * You can use this for exemple to differentiate data loaded form local app storage or a remote API
   */
  @observable
  public lastLoadedAt: PersistenceServiceName | boolean = false

  /**
   * The name of the scope, can be any string
   */
  public name: string

  /**
   * The parameters used to load the items
   * exemple : {status: 'active'}
   */
  public params: object

  /**
   * The scope's collection
   */
  public collection: Collection<RecordType>

  constructor(collection: Collection<RecordType>, name: string = "default", params: object = {}) {
    this.params = params
    this.name = name
    this.collection = collection
  }

  /**
   * Sets the scope's paramaters
   * @param {object} params
   * @return {this<RecordType extends Record>}
   */
  @action.bound
  public setParams(params: object) {
    this.params = params
    return this
  }

  /**
   * Load items for this scope into the scope's collection
   * @param {object} params
   * @return {Promise<any>}
   */
  @action.bound
  public load(params?: object): Promise<any> {
    return this.collection.load(params ? params : this.params, this.name)
  }

  /**
   * returns true if the scope contains the pk
   */
  public hasPk(pk: PrimaryKey) {
    return this.itemPrimaryKeys.findIndex(p => p == pk) !== -1
  }

  /**
   * adds primary key in the scope's pk list, if the latter is not already there
   */
  public addPk(pk: PrimaryKey) {
    if (!this.hasPk(pk)) {
      this.itemPrimaryKeys.push(pk)
    }

    return this
  }

  /**
   * replace primary key in the scope's pk list by another one
   */
  public replacePk(oldpk: PrimaryKey, newPk: PrimaryKey) {
    const pkIndex = this.itemPrimaryKeys.findIndex(p => p == oldpk)
    if (pkIndex !== -1) {
      this.itemPrimaryKeys[pkIndex] = newPk
    }
    return this
  }

  /**
   * Removes a pk form the set
   * @param {PrimaryKey} pk
   * @return {Scope<RecordType extends Record>}
   */
  public removePk(pk: PrimaryKey): Scope<RecordType> {
    const idx = this.itemPrimaryKeys.findIndex(p => p == pk)
    if (idx !== -1) {
      this.itemPrimaryKeys.splice(idx, 1)
    }
    return this
  }

  /**
   * get the scope items instances
   */
  @computed
  get items(): RecordType[] {
    const result = []

    for (let i = 0; i < this.itemPrimaryKeys.length; i++) {
      const item = this.collection.get(this.itemPrimaryKeys[i])
      if (item) {
        result.push(item)
      }
    }

    return result
  }
}
