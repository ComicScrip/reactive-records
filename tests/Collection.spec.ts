import { each } from "lodash"
import {
  Collection,
  Partial,
  PersistenceService,
  PersistenceStrategy
} from "../src/internals"
import { Album, AlbumCollection, albumCollection, ApiClient } from "./internals"
import { Scope } from "../src/Scope"
import { NetworkOnlyStrategy } from "./persistenceStrategies/NetworkOnlyStrategy"

describe("Collection", () => {
  beforeEach(() => {
    albumCollection.clear()
  })

  describe("#set", () => {
    it("can set a record from an object reprentation", () => {
      expect(albumCollection.size).toBe(0)
      const album = albumCollection.set({
        name: "my album",
        coverUrl: "the link"
      })
      expect(albumCollection.size).toBe(1)
      expect(album._ownAttributes).toMatchSnapshot()
    })

    it("does nothing if an exisiting record instance is passed as argument", () => {
      expect(albumCollection.size).toBe(0)
      const album = albumCollection.set({
        name: "my album",
        coverUrl: "the link"
      })
      albumCollection.set(album, false)
      expect(albumCollection.size).toBe(1)
      expect(album._ownAttributes).toMatchSnapshot()
    })

    it("throws an exception if record props are not valid", () => {
      expect(albumCollection.size).toBe(0)
      expect(() => {
        albumCollection.set({
          name: "my album",
          coverUrl: "the link",
          notADeclaredProp: 42
        } as Partial<Album>)
      }).toThrow()
    })
  })

  describe("#setMany", () => {
    it("can set records from an array of record plain objects", () => {
      expect(albumCollection.size).toBe(0)
      const albums = albumCollection.setMany([
        { name: "my album", coverUrl: "the link" },
        { name: "my second album" }
      ])
      expect(albumCollection.size).toBe(2)
      each(albums, album => expect(album._ownAttributes).toMatchSnapshot())
    })

    it("does not throw when and only when param 'strict' is set to false and undeclared record prop is given", () => {
      expect(albumCollection.size).toBe(0)
      expect(() => {
        albumCollection.setMany(
          [
            { name: "my album", coverUrl: "the link" },
            { name: "my second album", notDelared: "zdyzgf" } as Partial<Album>
          ],
          true
        )
      }).toThrow()

      expect(() => {
        albumCollection.setMany(
          [
            { name: "my album", coverUrl: "the link" },
            { name: "my second album", notDelared: "zdyzgf" } as Partial<Album>
          ],
          false
        )
      }).not.toThrow()
    })
  })

  describe("#itemsPrimaryKeys", () => {
    it("returns a list a the primary keys of the items", () => {
      const albums = albumCollection.setMany([
        { id: 1, name: "my album", coverUrl: "the link" },
        { id: 3, name: "my second album" }
      ])

      expect(albumCollection.itemsPrimaryKeys).toMatchSnapshot()
    })
  })

  describe("#items", () => {
    it("reflects the records present in _collection", () => {
      const albums = albumCollection.setMany([
        { name: "my album", coverUrl: "the link" },
        { name: "my second album" }
      ])

      expect(albumCollection.items[0]._ownAttributes).toMatchSnapshot()
      expect(albumCollection.items[1]._ownAttributes).toMatchSnapshot()
    })
  })

  describe("#has", () => {
    it("returns true if the provided primary key is present in the _collection and false otherwise", () => {
      const album = albumCollection.set({
        name: "my album",
        coverUrl: "the link"
      })
      expect(albumCollection.has(album._primaryKeyValue)).toEqual(true)
      expect(albumCollection.has("notexising")).toEqual(false)
    })
  })

  describe("#get", () => {
    it("can get a record from its primary key", () => {
      const album = albumCollection.set({
        name: "my album",
        coverUrl: "the link"
      })
      expect(albumCollection.get(album._primaryKeyValue)).toEqual(album)
    })
  })

  describe("#getMany", () => {
    it("can get records from an array of primary keys", () => {
      const albums = albumCollection.setMany([
        { name: "my album", coverUrl: "the link" },
        { name: "my second album" }
      ])

      const pks = albums.map(album => album._primaryKeyValue)
      expect(
        albumCollection.getMany(pks).map(a => a._ownAttributes)
      ).toMatchSnapshot()
      expect(
        albumCollection.getMany([pks[1]]).map(a => a._ownAttributes)
      ).toMatchSnapshot()
    })
  })

  describe("#unset", () => {
    it("can delete a record in the _collection from its primary key", () => {
      const album = albumCollection.set({
        name: "my album",
        coverUrl: "the link"
      })
      expect(albumCollection.size).toEqual(1)
      albumCollection.unset(album._primaryKeyValue)
      expect(albumCollection.size).toEqual(0)
    })
  })

  describe("#unsetMany", () => {
    it("can delete records from an array of primary keys", () => {
      const albums = albumCollection.setMany([
        { name: "my album", coverUrl: "the link" },
        { name: "my second album" },
        { name: "my third album" }
      ])

      expect(albumCollection.size).toEqual(3)
      const pks = albums.map(album => album._primaryKeyValue)
      albumCollection.unsetMany([pks[0], pks[1], "not_existing_pk"])
      expect(albumCollection.items[0]._ownAttributes).toMatchSnapshot()
    })
  })

  describe("scopes", () => {
    describe("#getScope", () => {
      it("should return the desired scope if it exists", () => {
        const ac = albumCollection as any
        const s = new Scope(albumCollection, "scope1")
        ac.scopes = new Map([["scope1", s]])
        expect((ac as AlbumCollection).getScope("scope1")).toBe(s)
      })

      it("should return undefined if the scope does not exist on the collection", () => {
        const ac = albumCollection as any
        const s = new Scope(albumCollection, "scope1")
        ac.scopes = new Map([["scope1", s]])
        expect((ac as AlbumCollection).getScope("scope2")).toBe(undefined)
      })
    })

    describe("provideScope", () => {
      it("should return the desired scope if it exists", () => {
        const ac = albumCollection as any
        const s = new Scope(albumCollection, "scope1")
        ac.scopes = new Map([["scope1", s]])
        expect((ac as AlbumCollection).provideScope("scope1")).toBe(s)
      })

      it("should return a new scope if the scope does not exist on the collection", () => {
        const ac = albumCollection as any
        const s = new Scope(albumCollection, "scope1")
        ac.scopes = new Map([["scope1", s]])
        const s2 = (ac as AlbumCollection).provideScope("scope2")
        expect(s2 instanceof Scope).toBe(true)
      })
    })
  })

  describe("persistence methods", () => {
    describe("load", () => {
      it("should call the collection's persistence stategy loadMany method with the right params", async () => {
        const loadAlbums = jest.fn()
        albumCollection.persistenceStrategy.loadMany = loadAlbums
        const defaultScope = albumCollection.provideScope()
        await albumCollection.load()
        expect(loadAlbums).toHaveBeenLastCalledWith({}, defaultScope)

        const customScope = albumCollection.provideScope("scope1")
        await albumCollection.load(customScope.name, { year: 1970 })
        expect(loadAlbums).toHaveBeenLastCalledWith({ year: 1970 }, customScope)
      })
    })

    describe("loadOne", () => {
      it("should call the collection's persistence stategy loadOne method with the right params", async () => {
        const loadAlbum = jest.fn()
        albumCollection.persistenceStrategy.loadOne = loadAlbum
        const defaultScope = albumCollection.provideScope()
        const record = albumCollection.set({ id: 2 })
        await albumCollection.loadOne(record)
        expect(loadAlbum).toHaveBeenLastCalledWith({}, record, defaultScope)

        const customScope = albumCollection.provideScope("scope1")
        await albumCollection.loadOne(record, customScope.name, { year: 1970 })
        expect(loadAlbum).toHaveBeenLastCalledWith(
          { year: 1970 },
          record,
          customScope
        )
      })
    })

    describe("saveOne", () => {
      it("should call the collection's persistence stategy saveOne method with the right params", async () => {
        const saveAlbum = jest.fn()
        albumCollection.persistenceStrategy.saveOne = saveAlbum
        const defaultScope = albumCollection.provideScope()
        const record = albumCollection.set({ id: 2 })
        await albumCollection.saveOne(record)
        expect(saveAlbum).toHaveBeenLastCalledWith({}, record, defaultScope)

        const customScope = albumCollection.provideScope("scope1")
        await albumCollection.saveOne(record, customScope.name)
        expect(saveAlbum).toHaveBeenLastCalledWith({}, record, customScope)
      })
    })

    describe("destroyOne", () => {
      it("should call the collection's persistence stategy destroyOne method with the right params", async () => {
        const destroyAlbum = jest.fn()
        albumCollection.persistenceStrategy.destroyOne = destroyAlbum
        const defaultScope = albumCollection.provideScope()
        const record = albumCollection.set({ id: 2 })
        await albumCollection.destroyOne(record)
        expect(destroyAlbum).toHaveBeenLastCalledWith({}, record, defaultScope)

        const customScope = albumCollection.provideScope("scope1")
        await albumCollection.destroyOne(record, customScope.name)
        expect(destroyAlbum).toHaveBeenLastCalledWith({}, record, customScope)
      })
    })
  })
})
