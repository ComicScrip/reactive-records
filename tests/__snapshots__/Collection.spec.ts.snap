// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collection #getMany can get records from an array of primary keys 1`] = `
Array [
  Object {
    "band_id": undefined,
    "coverUrl": "the link",
    "id": undefined,
    "name": "my album",
    "releaseDate": undefined,
  },
  Object {
    "band_id": undefined,
    "coverUrl": undefined,
    "id": undefined,
    "name": "my second album",
    "releaseDate": undefined,
  },
]
`;

exports[`Collection #getMany can get records from an array of primary keys 2`] = `
Array [
  Object {
    "band_id": undefined,
    "coverUrl": undefined,
    "id": undefined,
    "name": "my second album",
    "releaseDate": undefined,
  },
]
`;

exports[`Collection #items reflects the records present in _collection 1`] = `
Object {
  "band_id": undefined,
  "coverUrl": "the link",
  "id": undefined,
  "name": "my album",
  "releaseDate": undefined,
}
`;

exports[`Collection #items reflects the records present in _collection 2`] = `
Object {
  "band_id": undefined,
  "coverUrl": undefined,
  "id": undefined,
  "name": "my second album",
  "releaseDate": undefined,
}
`;

exports[`Collection #itemsPrimaryKeys returns a list a the primary keys of the items 1`] = `
Array [
  1,
  3,
]
`;

exports[`Collection #set can set a record from an object reprentation 1`] = `
Object {
  "band_id": undefined,
  "coverUrl": "the link",
  "id": undefined,
  "name": "my album",
  "releaseDate": undefined,
}
`;

exports[`Collection #set does nothing if an exisiting record instance is passed as argument 1`] = `
Object {
  "band_id": undefined,
  "coverUrl": "the link",
  "id": undefined,
  "name": "my album",
  "releaseDate": undefined,
}
`;

exports[`Collection #setMany can set records from an array of record plain objects 1`] = `
Object {
  "band_id": undefined,
  "coverUrl": "the link",
  "id": undefined,
  "name": "my album",
  "releaseDate": undefined,
}
`;

exports[`Collection #setMany can set records from an array of record plain objects 2`] = `
Object {
  "band_id": undefined,
  "coverUrl": undefined,
  "id": undefined,
  "name": "my second album",
  "releaseDate": undefined,
}
`;

exports[`Collection #setMany runs setmany in transaction 1`] = `
Object {
  "band_id": undefined,
  "coverUrl": "the link",
  "id": undefined,
  "name": "my album",
  "releaseDate": undefined,
}
`;

exports[`Collection #setMany runs setmany in transaction 2`] = `
Object {
  "band_id": undefined,
  "coverUrl": undefined,
  "id": undefined,
  "name": "my second album",
  "releaseDate": undefined,
}
`;

exports[`Collection #unsetMany can delete records from an array of primary keys 1`] = `
Object {
  "band_id": undefined,
  "coverUrl": undefined,
  "id": undefined,
  "name": "my third album",
  "releaseDate": undefined,
}
`;
